deepspeed/runtime/engine.py

```diff
@@ -1789,7 +1807,7 @@ def _take_model_step(self, lr_kwargs, block_eigenvalue={}):
         if (not self.zero_optimization() and not self.fp16_enabled()
                 and not self.amp_enabled()):
             self.zero_grad()
-        else:
+        elif not self.bfloat16_enabled():
```

Comment:
Do we not zero grad when bf16 is enabled?

Reply:
bf16 manages its own grads because remember it accumulates them in a special fp32 buffer which gets cleared after step, see:

https://github.com/microsoft/DeepSpeed/pull/1801/files#diff-e6635a81d2c2bf0938b5f83b1c4945e0f344e0106191a6960df8ee4aa64cc55fR1810

Reply:
but I'm actually not sure if the branch above is correct, since it'd be true for bf16 I think - let me check in debugger.

Reply:
it totally goes into line 1809 as I suspected.

@tjruwase, is that by design - the `elif` on 1810 will never get executed under bf16.

Reply:
we rewrote that logic as it was somewhat broken, but wasn't really a bug.

Reply:
https://github.com/microsoft/DeepSpeed/pull/1801/commits/144f652768a980b7a4f1e1104ae7872189742c99